### PR TITLE
#3233 UI: Common method to delete server and custom directory entries

### PR DESCRIPTION
**Pull request by AdamGLIN**
I added a horizontal box layout which include the box and the delete button. I wrote a fonction to implement the delete button.

It works well when we change the focus with the mouse to add a new directory but I can't make the enter key work : it delete the input and nothing happend. I tried to add the QLineEdit::returnPressed signal but it changed nothing. I struggle to understand the former use of the QComboBox::activated signal since I commented it and nothing seems to have changed.

TODO :
Allowing to add directory via enter key.
Improve spacers in the ui.

**Short description of changes**

CHANGELOG: UI: Add "Delete Entry" button to Advanced Settings, Custom Directories

**Context: Fixes an issue?**

#3233 

**Does this change need documentation? What needs to be documented and how?**

**Status of this Pull Request**

**What is missing until this pull request can be merged?**

## Checklist

-  [x] I've verified that this Pull Request follows the [general code principles](https://github.com/jamulussoftware/jamulus/blob/main/CONTRIBUTING.md#jamulus-projectsource-code-general-principles)
-  [x] I tested my code and it does what I want
-  [x] My code follows the [style guide](https://github.com/jamulussoftware/jamulus/blob/main/CONTRIBUTING.md#source-code-consistency) 
-  [x] I waited some time after this Pull Request was opened and all GitHub checks completed without errors. 
-  [x] I've filled all the content above

### CODE REVIEW

**Comment by ann0see** [comment]
Thanks for opening this PR. It's a draft/WIP so it needs discussion. @pljones Should be most familiar with this.

**Comment by ann0see** [comment]
Could you please also post your questions here? Best right in the code tab close to where the question arises.

**Comment by softins** [comment]
Thanks Adam! I started to look at your commit and questions earlier today, but got diverted onto other things. I will have a look soon, hopefully tomorrow!

File: src/clientsettingsdlg.cpp [hunk 1]
```diff
@@ -360,14 +360,17 @@ CClientSettingsDlg::CClientSettingsDlg ( CClient* pNCliP, CClientSettings* pNSet
...
-                                        "then move focus out of the control." );
+                                        "you can enter the addresses here.<br>" );
 
     lblCustomDirectories->setWhatsThis ( strCustomDirectories );
     cbxCustomDirectories->setWhatsThis ( strCustomDirectories );
     cbxCustomDirectories->setAccessibleName ( tr ( "Custom Directories combo box" ) );
 
+    butDeleteCustomDirectories->setAccessibleName ( tr ( "Delete custom directory button" ) );
+    butDeleteCustomDirectories->setWhatsThis ( "<b>" + tr ( "Delete Custom Directory" ) + ":</b> " +
+                                               tr ( "Click the button to clear the currently selected custom directory "
```
Comment (pljones) [inline src/clientsettingsdlg.cpp:371]: There's a specific usage difference between the Connect dialog and the Settings dialog.

On the Connect dialog, if no server is selected in the server list and the server combo box field is empty, hitting connect will default to the server hosting the current directory.

As you'll understand, that makes no sense for the Settings screen.  So the phrasing of the "What's This" needs to be a bit simpler: just refer to deleting the entry.

File: src/clientsettingsdlg.cpp [hunk 2]
```diff
@@ -691,12 +694,16 @@ CClientSettingsDlg::CClientSettingsDlg ( CClient* pNCliP, CClientSettings* pNSet
                        this,
                        &CClientSettingsDlg::OnMeterStyleActivated );
 
+    QObject::connect ( butDeleteCustomDirectories, &QPushButton::clicked, this, &CClientSettingsDlg::OnDeleteCustomDirectoriesClicked );
```
Comment (pljones) [inline src/clientsettingsdlg.cpp:697]: This new field surely comes after the combobox, so the code here should, logically, come after the connects fpr the combobox.

File: src/clientsettingsdlg.cpp [hunk 3]
```diff
@@ -1009,17 +1016,37 @@ void CClientSettingsDlg::OnEnableOPUS64StateChanged ( int value )
 
 void CClientSettingsDlg::OnFeedbackDetectionChanged ( int value ) { pSettings->bEnableFeedbackDetection = value == Qt::Checked; }
 
-void CClientSettingsDlg::OnCustomDirectoriesEditingFinished()
+void CClientSettingsDlg::OnDeleteCustomDirectoriesClicked()
 {
-    if ( cbxCustomDirectories->currentText().isEmpty() && cbxCustomDirectories->currentData().isValid() )
+    if ( cbxCustomDirectories->currentText().isEmpty() )
```
Comment (pljones) [inline src/clientsettingsdlg.cpp:1021]: The ` && cbxCustomDirectories->currentData().isValid()` check is probably still worth having.

File: src/clientsettingsdlg.cpp [hunk 4]
```diff
@@ -1009,17 +1016,37 @@ void CClientSettingsDlg::OnEnableOPUS64StateChanged ( int value )
 
 void CClientSettingsDlg::OnFeedbackDetectionChanged ( int value ) { pSettings->bEnableFeedbackDetection = value == Qt::Checked; }
 
-void CClientSettingsDlg::OnCustomDirectoriesEditingFinished()
+void CClientSettingsDlg::OnDeleteCustomDirectoriesClicked()
```
[the remark this answers is not in the document]
Comment (pljones) [reply src/clientsettingsdlg.cpp:1019]: (In fact, both signal handlers could call a common method passing an argument to indicate "delete" or "add", so that the only change would be working out whether an entry was being deleted by seeing if the field is empty.  Adding would have to prevent adding empty values, of course.)

File: src/clientsettingsdlgbase.ui [hunk 5]
```diff
@@ -1038,13 +1038,41 @@
...
+             <layout class="QHBoxLayout">
              <item>
               <widget class="QComboBox" name="cbxCustomDirectories">
                <property name="editable">
                 <bool>true</bool>
                </property>
               </widget>
              </item>
+             <item>
+             <widget class="QPushButton" name="butDeleteCustomDirectories">
```
Comment (pljones) [inline src/clientsettingsdlgbase.ui:1051]: The button only deletes a single entry so the name should be singular.

File: src/clientsettingsdlg.cpp (hunk 4 above)
Comment (pljones) [inline src/clientsettingsdlg.cpp:1019]: This should be a one line wrapper around the common code, passing in a "delete current entry" flag.

File: src/clientsettingsdlg.cpp [hunk 6]
```diff
@@ -1009,17 +1016,37 @@ void CClientSettingsDlg::OnEnableOPUS64StateChanged ( int value )
...
+    // empty last entry
+    pSettings->vstrDirectoryAddress[MAX_NUM_SERVER_ADDR_ITEMS - 1] = QString();
+
+    // redisplay to pick up updated list
+    UpdateDisplay();
+    UpdateDirectoryComboBox();
+    emit CustomDirectoriesChanged();
+}
+
+void CClientSettingsDlg::OnCustomDirectoriesEditingFinished()
```
Comment (pljones) [inline src/clientsettingsdlg.cpp:1046]: This should be a one line wrapper around the current code, passing in an "update current entry" flag.

File: src/clientsettingsdlg.cpp [hunk 7]
```diff
@@ -1009,17 +1016,37 @@ void CClientSettingsDlg::OnEnableOPUS64StateChanged ( int value )
 
 void CClientSettingsDlg::OnFeedbackDetectionChanged ( int value ) { pSettings->bEnableFeedbackDetection = value == Qt::Checked; }
 
-void CClientSettingsDlg::OnCustomDirectoriesEditingFinished()
+void CClientSettingsDlg::OnDeleteCustomDirectoriesClicked()
 {
     if ( cbxCustomDirectories->currentText().isEmpty() && cbxCustomDirectories->currentData().isValid() )
```
Comment (pljones) [inline src/clientsettingsdlg.cpp:1021]: This needs to change to check whether the call was made for "delete current entry".

If it was, it then needs the existing check that current entry is valid and then it can delete the entry.  The old line 1019 should be all that's needed to do that.  It shouldn't need the `for ... while ...` loops.

It also doesn't need to return.  The `else` (i.e. "not the delete button") case will prevent anything unintended happening.

File: src/clientsettingsdlg.cpp [hunk 8]
```diff
@@ -1009,17 +1016,37 @@ void CClientSettingsDlg::OnEnableOPUS64StateChanged ( int value )
...
+    {
+        while ( pSettings->vstrDirectoryAddress[iLEIdx].compare ( NetworkUtil::FixAddress ( cbxCustomDirectories->currentText() ) ) == 0 )
+        {
+            for ( int jLEIdx = iLEIdx + 1; jLEIdx < MAX_NUM_SERVER_ADDR_ITEMS; jLEIdx++ )
+            {
+                pSettings->vstrDirectoryAddress[jLEIdx - 1] = pSettings->vstrDirectoryAddress[jLEIdx];
+            }
+        }
     }
-    else if ( cbxCustomDirectories->currentData().isValid() && pSettings->vstrDirectoryAddress[cbxCustomDirectories->currentData().toInt()].compare (
```
Comment (pljones) [inline src/clientsettingsdlg.cpp:1021]: The `else` here will be for "not the delete button".  The whole of the `if ... else` block that follows should be inside that `else`.

So "if they didn't click the delete button, we got here because editing finished".  Check the data is valid and do nothing if nothing changed.  Else update the entry.

File: src/clientsettingsdlg.cpp [hunk 9]
```diff
@@ -1009,17 +1016,37 @@ void CClientSettingsDlg::OnEnableOPUS64StateChanged ( int value )
...
+    emit CustomDirectoriesChanged();
+}
+
+void CClientSettingsDlg::OnCustomDirectoriesEditingFinished()
+{
+    if ( cbxCustomDirectories->currentData().isValid() && pSettings->vstrDirectoryAddress[cbxCustomDirectories->currentData().toInt()].compare (
+                                                              NetworkUtil::FixAddress ( cbxCustomDirectories->currentText() ) ) == 0 )
     {
         // if the user has selected another entry in the combo box list without changing anything,
         // there is no need to update any list
```
Comment (pljones) [inline src/clientsettingsdlg.cpp:1052]: At (old) line 1035, the outer "if delete or not" block should have closed and the method can continue to update the dialog box and emit the signal.

File: src/clientsettingsdlg.cpp [hunk 10]
```diff
@@ -1009,29 +1010,19 @@ void CClientSettingsDlg::OnEnableOPUS64StateChanged ( int value )
 
 void CClientSettingsDlg::OnFeedbackDetectionChanged ( int value ) { pSettings->bEnableFeedbackDetection = value == Qt::Checked; }
 
-void CClientSettingsDlg::OnCustomDirectoriesEditingFinished()
+void CClientSettingsDlg::OnCustomDirectoriesChanged ( bool bDelete )
 {
```
Comment (pljones) [inline src/clientsettingsdlg.cpp:1014]: `cbxCustomDirectories->currentData().isValid()` looks common to both legs of the `if` -- so the method could just return (doing nothing) if it's false.

File: src/clientsettingsdlg.cpp [hunk 11]
```diff
@@ -1009,29 +1010,19 @@ void CClientSettingsDlg::OnEnableOPUS64StateChanged ( int value )
...
     }
-    else if ( cbxCustomDirectories->currentData().isValid() && pSettings->vstrDirectoryAddress[cbxCustomDirectories->currentData().toInt()].compare (
-                                                                   NetworkUtil::FixAddress ( cbxCustomDirectories->currentText() ) ) == 0 )
-    {
-        // if the user has selected another entry in the combo box list without changing anything,
-        // there is no need to update any list
-        return;
-    }
-    else
+    else if ( !cbxCustomDirectories->currentText().isEmpty() )
```
Comment (pljones) [inline src/clientsettingsdlg.cpp:1020]: As above - still needs to be _valid_ - otherwise the index isn't correct for the delete.

Comment (pljones) [inline src/clientsettingsdlg.cpp:1020]: The old test is the right one:
```suggestion
    if ( cbxCustomDirectories->currentData().isValid() && pSettings->vstrDirectoryAddress[cbxCustomDirectories->currentData().toInt()].compare ( cbxCustomDirectories->currentText() ) ) == 0 )
```

File: src/clientsettingsdlg.cpp [hunk 12]
```diff
@@ -1009,29 +1010,19 @@ void CClientSettingsDlg::OnEnableOPUS64StateChanged ( int value )
 
 void CClientSettingsDlg::OnFeedbackDetectionChanged ( int value ) { pSettings->bEnableFeedbackDetection = value == Qt::Checked; }
 
-void CClientSettingsDlg::OnCustomDirectoriesEditingFinished()
+void CClientSettingsDlg::OnCustomDirectoriesChanged ( bool bDelete )
 {
-    if ( cbxCustomDirectories->currentText().isEmpty() && cbxCustomDirectories->currentData().isValid() )
+    if ( bDelete && cbxCustomDirectories->currentData().isValid() )
```
Comment (pljones) [inline src/clientsettingsdlg.cpp:1015]: This should be:
```pseudocode
IF the user clicked delete
    IF the entry is not valid THEN return ENDIF
    delete the entry
ELSE ...

File: src/clientsettingsdlgbase.ui [hunk 13]
```diff
@@ -1031,19 +1031,48 @@
              <item>
               <widget class="QLabel" name="lblCustomDirectories">
                <property name="text">
-                <string>Custom Directories:</string>
+                <string>Custom Directories</string>
```
Comment (pljones) [inline src/clientsettingsdlgbase.ui:1034]: Why this change?  This affects translations.

**Comment by pljones** [comment]
Looks good to me, now.  I've kicked off the build.

**Comment by AdamGLIN** [comment]
Thank you for your patience and your advices :)

**Comment by pljones** [comment]
Just tried the Windows version from the build artifacts and it looks okay to me.

**Review by pljones — APPROVED** [review]

**Comment by ann0see** [comment]
This needs some squashing and rebasing, I believe. I'll look into it once I have time.

**Review by softins — APPROVED** [review]
I've tried this on my Pi and my Mac. Both ok, so happy to approve.

There don't seem to be any merge conflicts, so a squash merge could be done without needing to rebase. Or else I'd be happy to rebase and squash manually first.

**Comment by pljones** [comment]
I prefer everything neat and tidy, so if you could rebase and squash (happy to do it, too).

**Comment by softins** [comment]
OK, I'll rebase it onto main, which now has the Mac legacy Qt update, let it build, and then check it on my Mac.

**Review by softins — CHANGES_REQUESTED** [review]
I've just loaded the Mac legacy version built with Qt 5.15 and noticed a couple of issues which need to be addressed first. I'm happy to look at these.

1. The button is too narrow for the delete character (it's ok on other platforms and was ok with Qt 5.9). This is the same as I reported in #3241 
image

2. If I type in a custom directory address and finish with "Enter", the field goes blank and the address isn't added. I need to use Tab, or to click outside the field, to make it lose focus. I did see some comments about pressing Enter, but thought they had been resolved. Evidently not, at least on some platforms.

**Comment by pljones** [comment]
Yep Qt on MacOS certainly seems to come with its own UI layout peculiarities that aren't present on other platforms.

The action/focus trigger for updating the list has also been a bit hard to pin down: it works _right_ sometimes on both Windows and Linux for me.  It also works _wrong_ sometimes, too...

**Comment by softins** [comment]
I fixed the issue with Enter. The problem was that pressing Enter within the combi box was causing a button push to delete, because the default setting for `autoDefault` appears to be `true`. Having set `autoDefault` explicitly to `false` for the delete button in `clientsettingsdlgbase.ui`, the Enter key works correctly for just finishing editing.

Will investigate the button appearance on Mac later.

**Comment by softins** [comment]
For the delete buttons, I needed to change the horizontal size policy from "Maximum" to "Fixed", so that the buttons were definitely wide enough, but not allowed to grow. Need to check they are ok on all platforms.

**Comment by softins** [comment]
I've tried the latest build above on Mac Legacy, Windows and Pi Linux. Not current Mac though.

The button width is now fine on Mac Legacy, but it's a bit wider than ideal on Windows and Linux. Must be scaling differences. It may be acceptable, but if not, we would need to add code to adjust the width of those buttons at runtime for Mac only. Thoughts?

**Review by softins — APPROVED** [review]

**Comment by pljones** [comment]
Maybe set the size policy to maximum only for Mac if that's where it's needed?

**Comment by softins** [comment]
> Maybe set the size policy to maximum only for Mac if that's where it's needed?

It's the other way around. It was previously maximum only, but that allows the button to be shrunk as much as needed. But on Mac from 5.12 on, the button was shrinking right down. My guess is that it couldn't correctly calculate the size of the contained text when it was just a Unicode character for backspace, so took the text width as 0. I found I needed to specify Fixed so that both maximum and minimum were constrained.

Either way, to fix it we would need to add conditional runtime code for Mac only, as there is no way to put architecture conditionals in a `.ui` file. The question is, is it worth doing?

I have wondered whether it might be worth putting an image in the button instead of a Unicode character? e.g. a wastebasket or trashcan?

**Comment by softins** [comment]
A further thought: maybe merge this PR without the button width fix, since that applies to both the settings and connect dialogs? The connect dialog is arguably outside the scope of this PR. Then raise a separate PR to fix #3241 in both places for Mac?

**Comment by pljones** [comment]
OK by me.

**Review by pljones — APPROVED** [review]

**Comment by pljones** [comment]
image
image
Hmm... they're not identical.  The button in the server is larger and the glyph in the client is bold...

**Comment by ann0see** [comment]
Ok. So this also needs new screenshots for the docs potentially.

**Comment by pljones** [comment]
Linux:
image

Windows:
image

**Comment by ann0see** [comment]
Yes. But that should show up

**Comment by pljones** [comment]
https://github.com/jamulussoftware/jamuluswebsite/issues/1002 raised for documentation.
